### ISSUE TITLE
Apparent resource leak in jemalloc 5.1.0 running on 64-bit windows

### ISSUE THREAD

**Issue by Cytoo** [jemalloc/jemalloc#1263]
Hello ! When trying to update our allocator from jemalloc 4.3.1 to version 5.1.0, we encounter a severe resource leak that causes our application to crash. This might be related to issue #1152 

Conditions : Our application is a server that handles requests. We have a stress suite that simulates request with heavy processing on server side, in order to test our application stability.

Setup : 36 physical cores, 128 GB of memory.

With Jemalloc 4.3.1 :  After running the suite, the server process takes 2.9GB of Working set and 2.6GB of Commit size under windows task manager.

With Jemalloc 5.1.0: After running the suite, the server process takes 21.6GB of Working set and 25.4GB of Commit size. It keeps increasing if we launch the suite again, until OOM occurs. The server is supposed to release most of acquired resources after the suite, (leaks are not excluded though), but with 4.3.1 it is mostly stable after the run.

Here are both malloc stats:
Please note that I fiddle a little bit with jemalloc 5 options first to see if it was faulty.

[jemalloc5.txt](https://github.com/jemalloc/jemalloc/files/2145387/jemalloc5.txt)

[jemalloc4.3.txt](https://github.com/jemalloc/jemalloc/files/2145388/jemalloc4.3.txt)

**Comment by interwq** [comment]
Thanks for sharing the info! The stats with 5.1 do show very high mapped memory (~25G which is 10x of active memory ~2G). Can you please help us to try a couple more options to identify the problem? Adding "muzzy_decay_ms:0" and "retain:true" to malloc_conf, also disable fill (configure with --disable-fill) -- if they fix the issue maybe try them separately to see which one is the cause?

**Comment by interwq** [comment]
It would also help if you can share the repro program and steps with us.

**Comment by Cytoo** [comment]
muzzy_decay_ms:0, retain:true and disable-fill did not change much the behaviour, except now it takes a bit more of memory after the run (23.3GB working set, 29.4GB commit size).

I can't share the reproducing program though.

**Comment by Cytoo** [comment]
I modified the test_threads file in order to allocate/free in other threads and observe a leak that is not there when testing in 4.3.1

[test_threads.txt](https://github.com/jemalloc/jemalloc/files/2150114/test_threads.txt)

Also, the value of allocated seems to be wrong only after the call to malloc_stats but i figured this might be because the value of stats.active is not refreshed otherwise ?

I will start a git bisect from 4.3.1 to try and detect: 

1. The leak then OOM in our as observed when running our server
2. The leak in test_threads

**Comment by Cytoo** [comment]
Update, because the test_threads.txt file is not the good file !

[test_threads.txt](https://github.com/jemalloc/jemalloc/files/2154362/test_threads.txt)

This one is the proper file where we witness a leak

**Comment by Cytoo** [comment]
Bisecting was quite hard because a lot of bugs interacted together so i could not quite separate causes. And besides, two leaks may look the same and yet have different causes.

I knew version tagged 4.3.1 was OK, because this is the version we are currently using in our product. So I bisected between dev and 4.3.1

Commit identified as introducing the bug is commit de0305a7f3e443d48e012272e1b91c44d2b129d2 but could not be reproduced using the test_threads program, I am guessing because there is a later commit introducing the change that made print_stats aware of the leak.
I am not 100% sure because at this time in the history there were multiple bugs:
- Infinite recursion causing stack overflow
- bad_alloc exception
- the leak I am looking for

Nonetheless, commit de0305a7f3e443d48e012272e1b91c44d2b129d2 changes were introduced to master branch in merge commit "5.0.0" 5018fe3f0979b7f9db9930accdf7ee31071fd703

**Comment by Cytoo** [comment]
I did another bisect to try and find the commit that allows test_threads to detect the leak :  fde3e20cc04db459f3c76134bc6dfb0ee5c422bb

Before that, the leak was transparent to test_thread

**Comment by interwq** [comment]
Thanks for all the details @Cytoo! I’m on vacation this week. I’ll look into this next week.

**Comment by rustyx** [comment]
@Cytoo you should flush the tcache of the main thread because you're freeing stuff in it.

Do this before collecting the final statistics:

```
  je_mallctl("thread.tcache.flush", nullptr, nullptr, nullptr, 0);
```

With that change the program you've [mentioned](https://github.com/jemalloc/jemalloc/issues/1263#issuecomment-401706502) shows 0 leaks.

The tcache of worker threads is cleaned automatically when they finish, and the main thread never allocated or freed anything in the [original](https://github.com/jemalloc/jemalloc/blob/33df2fa1694c9fdc1912aecaa19babc194f377ac/msvc/test_threads/test_threads.cpp) test_threads.cpp, that's why it never had to flush the tcache.

**Comment by santagada** [comment]
I think this is the same leak I have when moving from jemalloc 4.4 to 5.1 described in #1335

**Comment by interwq** [comment]
This should be fixed by the 5.2.1 release -- https://github.com/jemalloc/jemalloc/releases/tag/5.2.1
